File: java/src/com/esielkar/composite/Composite.java
package com.esielkar.composite;

import java.util.List;

public class Composite implements Component {
    private static int autoId = 0;

    private int id = ++autoId;

    private List<Component> children;

    public Composite(List<Component> children) {
        this.children = children;
    }

    public void add(Component component) {
        children.add(component);
    }

    public void remove(Component component) {
        children.remove(component);
    }

    public List<Component> getChildren() {
        return children;
    }

    @Override
    public void execute() {
        System.out.println("Composite " + id);
        for (Component component : children) {
            component.execute();
        }

    }

}

File: dart/design_patterns/bin/composite/composite.dart
import 'package:meta/meta.dart';

abstract class Component {
  void execute();
}

@sealed
class Leaf extends Component {
  static int _sid = 0;
  final id = ++_sid;
  @override
  void execute() {
    print('Leaf $id - Does work');
  }
}

class Composite extends Component {
  static int _sid = 0;
  final id = ++_sid;
  final List<Component> _children;

  Composite([List<Component>? children])
      : _children = children ?? <Component>[];

  List<Component> get children => _children.toList(growable: false);
  void add(Component component) => _children.add(component);
  bool remove(Component component) => _children.remove(component);

  @override
  void execute() {
    print('Composite $id');
    for (Component component in _children) {
      component.execute();
    }
  }
}
